### PR TITLE
[Translator] Unify EmitPackedHalvingAddS{8,16} implementations.

### PR DESCRIPTION
**Pull request by inquisitio**

### CODE REVIEW

**Comment by inquisitio** [comment]
hmm, I'm not able to reproduce the failure locally. Any ideas about what went wrong?

**Comment by FernandoS27** [comment]
It does not seem to pass the tests, you should check it out before submitting them as PR.

**Comment by inquisitio** [comment]
Thanks Fernando. Is there any way to get detailed output (or at least an understanding of why the test is failing?) The output seems to be pretty bare.

**Comment by inquisitio** [comment]
And sorry for breaking this, I expected this not to be a functional change.

**Comment by FernandoS27** [comment]
are you sure you are building and running dynarmic_tests ? it prints all the info on a mismatch. The registers, instruction, intial values, later values, interpreter's result and dynarmic's.

File: src/backend_x64/emit_x64.cpp [hunk 1]
```diff
@@ -1834,37 +1835,17 @@ void EmitX64::EmitPackedHalvingAddS8(RegAlloc& reg_alloc, IR::Block&, IR::Inst*
     code->mov(carry, xor_a_b);
     code->and(carry, 0x80808080);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: This constant (`0x80808080`) is not the same between both.

The masks are related; they are binary inverses.

File: src/backend_x64/emit_x64.cpp [hunk 2]
```diff
@@ -1834,37 +1835,17 @@ void EmitX64::EmitPackedHalvingAddS8(RegAlloc& reg_alloc, IR::Block&, IR::Inst*
...
-    Xbyak::Reg32 result = reg_a;
-    Xbyak::Reg32 carry = reg_alloc.ScratchGpr().cvt32();
-
-    // This relies on the equality x+y == ((x&y) << 1) + (x^y).
-    // Note that x^y always contains the LSB of the result.
-    // Since we want to calculate (x+y)/2, we can instead calculate (x&y) + ((x^y)>>1).
-    // We mask by 0x7FFF to remove the LSB so that it doesn't leak into the field below.
-    // carry propagates the sign bit from (x^y)>>1 upwards by one.
+void EmitX64::EmitPackedHalvingAddS8(RegAlloc& reg_alloc, IR::Block&, IR::Inst* inst) {
+    return EmitPackedHalvingAddSHelper(reg_alloc, inst, code, 0x7F7F7F7F);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: Unnecessary `return` staement.

File: src/backend_x64/emit_x64.cpp [hunk 3]
```diff
@@ -1834,37 +1835,17 @@ void EmitX64::EmitPackedHalvingAddS8(RegAlloc& reg_alloc, IR::Block&, IR::Inst*
...
-    code->and(and_a_b, reg_b);
-    code->xor(xor_a_b, reg_b);
-    code->mov(carry, xor_a_b);
-    code->and(carry, 0x80008000);
-    code->shr(xor_a_b, 1);
-    code->and(xor_a_b, 0x7FFF7FFF);
-    code->add(result, xor_a_b);
-    code->xor(result, carry);
+void EmitX64::EmitPackedHalvingAddS16(RegAlloc& reg_alloc, IR::Block&, IR::Inst* inst) {
+    return EmitPackedHalvingAddSHelper(reg_alloc, inst, code, 0x7FFF7FFF);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: Unnecessary `return` staement.

File: src/backend_x64/emit_x64.cpp [hunk 4]
```diff
@@ -1811,7 +1811,8 @@ void EmitX64::EmitPackedHalvingAddU16(RegAlloc& reg_alloc, IR::Block&, IR::Inst*
     code->add(result, xor_a_b);
 }
 
-void EmitX64::EmitPackedHalvingAddS8(RegAlloc& reg_alloc, IR::Block&, IR::Inst* inst) {
+static void EmitPackedHalvingAddSHelper(RegAlloc& reg_alloc, IR::Inst* inst,
+                                        BlockOfCode* code, u32 mask) {
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: Rename `mask` to `field_mask`, since that's a more descriptive name.

**Comment by MerryMage** [comment]
Closing due to inactivity.

**Pull request closed without merging**